### ISSUE TITLE
Is there a way to detect collisions?

### ISSUE THREAD

**Issue by aqvia** [esmini/esmini#58]
Like Collision detector on Carla.
It seems that the positions and velocities of entities for each timestamp can be seen using CSV log or osireceiver, but I don't know if the detection information can be obtained when entities collides (or overlaps).

**Comment by eknabevcc** [comment]
Well, I admit it would be nice to have triggers and actions within the CSV log. Something we should think about. 
But one way is to find it in the log file (or console output). Since release 2.0.7 time stamp is added to each log entry. 
Example (pedestrian_collision.xosc):
```
0.259 OSCAction.cpp / 96 / scenarioengine::StoryBoardElement::SetState(): ped_walk_event standbyState -> startTransition -> runningState
1.022 OSCAction.hpp / 129 / scenarioengine::StoryBoardElement::End(): walk_speed complete after 1 execution
1.041 OSCAction.cpp / 96 / scenarioengine::StoryBoardElement::SetState(): walk_speed runningState -> endTransition -> completeState
5.374 OSCCondition.cpp / 711 / scenarioengine::TrigByCollision::CheckCondition(): ped_fall_condition == true, dist: 4.85, edge: Rising
5.374 OSCAction.hpp / 129 / scenarioengine::StoryBoardElement::End(): walk_route complete after 1 execution
5.374 OSCAction.hpp / 129 / scenarioengine::StoryBoardElement::End(): ped_walk_event complete after 1 execution
```
At time = 5.374 you find the collision happening.

Note that you can specify fixed timestep  (e.g. --fixed_timestep 0.01) if you want to control time resolution.

Also note that the current implementation of collision detection is very approximative, something like checking if reference positions are within bounding box length of the triggering object.

**Comment by aqvia** [comment]
@eknabevcc 
Thank you for your reply.
So it means that, by setting CollisionCondition for target entities to StartTrigger and setting them to stop when a collision occurs, I can detect whether entities collide in the simulation. Right?

**Comment by eknabevcc** [comment]
I now suddenly better understand your need. You want to have a generic monitoring of any collision involving Ego vehicle. That makes sense. Perhaps could be done with CollisionCondition. But I think it's a bit cumbersome.

Another approach would be for esmini to detect all Ego-collisions and report it separately somehow. But that is not implemented yet.

But a third approach, which is already possible, would be to read out all object states (with SE_GetObjectState), which includes position, rotation and bounding box. And then "simply" calculate collisions with method of choice (depending on precision need). The simplest approach would be to measure Euclidean distance between reference points. Next level is taking bound box into consideration.

Right now we have some other higher prio stuff in backlog. But collision detection is something that esmini should eventually provide. Hope you can find a short term solution until then.

**Comment by eknabevcc** [comment]
Actually, I found your use case interesting. I added support for ByType option, and improved the logging of triggered conditions. So it's now possible (but maybe not optimal) to use CollisionCondition for detecting collisions between a vehicle, e.g. Ego, and all other of specified type. 
Implemented in commit 028db3f. Will be part of next release.
I attach a simple example scenario with interactive (arrow keys) Ego to illustrate collision detection. (had to rename to .txt in order to attach - you can rename to .xosc).
[dummy_collision_detection.txt](https://github.com/esmini/esmini/files/5654004/dummy_collision_detection.txt)

**Comment by eknabevcc** [comment]
Due to inactivity I close this for now. Please re-open if you have more questions or issues.

**Comment by eknabevcc** [comment]
By the way, just wanted to let you know about improved collision detection - now based on bounding box overlap.
Added in commit be2ef8a and improved in commit 2e22477. Part of [releases](https://github.com/esmini/esmini/releases/latest) since v2.1.12.

**Comment by eknabevcc** [comment]
Oops, wrong links. Should be: Added in commit 2e22477 and improved in commit dac9bf9.

**Comment by eknabevcc** [comment]
Global collision detection implemented by commit 10ccab03996fcaa05edc2702fae7e12021707cfe and included in release [2.19.1](https://github.com/esmini/esmini/releases/tag/v2.19.0)
